executor/show.go

```diff
@@ -1011,9 +1011,9 @@ func ConstructResultOfShowCreateTable(ctx sessionctx.Context, tableInfo *model.T
 	buf.WriteString(") ENGINE=InnoDB")
 	// We need to explicitly set the default charset and collation
 	// to make it work on MySQL server which has default collate utf8_general_ci.
-	if len(tblCollate) == 0 || tblCollate == "binary" {
+	if len(tblCollate) == 0 || (!mysql.IsUTF8Charset(tblCharset) && getDefaultCollate(tblCharset) == tblCollate) {
```

Comment:
To be honest, I don't think this PR should be merged. Keep the `collate` in `show create table` is better. Because MySQL will change the default collate in someday.